Ability to save and load field mapping

paustint opened the pull request:
Allow users to save and load field mapping

DEMO: https://jam.dev/c/4fd5541b-6191-41d0-8499-3b0e0c069417

![image](https://user-images.githubusercontent.com/5461649/234435587-9e89fe96-6065-434f-81ed-45cabf63d6e6.png)


![image](https://user-images.githubusercontent.com/5461649/234435606-574b25da-6455-4ab5-a419-063948f8ed37.png)


resolves #303

joiecosby commented on apps/jetstream/src/app/components/load-records/components/LoadRecordsFieldMappingRow.tsx:
```diff
@@ -45,40 +77,11 @@ export const LoadRecordsFieldMappingRow: FunctionComponent<LoadRecordsFieldMappi
-        ),
-      }))
-    );
+  useNonInitialEffect(() => {
```
Out of curiosity, can you explain why you moved this from useEffect() to useNonInitialEffect()? What is the benefit of this not running when it is initially rendered?

paustint replied:
@joiecosby -The main reason here is that `getFieldListItems(fields)` is sort of expensive since it has to iterate through every item and also creates some components.

Since `const [fieldListItems, setFieldListItems] = useState<ListItem<string, FieldWithRelatedEntities>[]>(() => getFieldListItems(fields));` already did the expensive work on the very first render, if we used a normal `useEffect` that would just mean that we have to do the exact same work twice for no good reason other than that is how react works.


I will send you a slack message with an excersize :D

joiecosby commented:
Another build question, I see that  LoadMappingPopover is the Load button in the UI and SaveMappingItem is the layout once you click the Load button/a child component. Is it best to granulize as much as possible? I guess I am curious if I should  have the mindset to create child components if I start doing more than the intentded functionality within a class.

joiecosby reviewed (approved):
LGTM!

paustint commented:
@joiecosby 
> Another build question, I see that LoadMappingPopover is the Load button in the UI and SaveMappingItem is the layout once you click the Load button/a child component. Is it best to granulize as much as possible? I guess I am curious if I should have the mindset to create child components if I start doing more than the intended functionality within a class.

There is no science behind it, I just figure it out as I go. Here is what I normally do
* I usually have a good idea up-front if what I want to start needs a new component, so that is the baseline
* 👉  I start by putting everything in one component so I can quickly work through getting something created
  * Just start with this in 90% of cases and then you can easily refactor and split up
* Once I have a better idea of the feature, I then usually know where it might be a good place to start splitting the code up
  * Do I have a lot of complicated logic in the component? (if pure computation logic, then maybe a custom hook could be good to encapsulate that)
  * Am I iterating through a list of items and have a bunch of crazy things using indexes ex.. to access data (if yes - FOR SURE put this in a child component - this is the `#1` thing that can simplify stuff
* Otherwise I split up if it feels right or if the component starts getting complicated or bloated
